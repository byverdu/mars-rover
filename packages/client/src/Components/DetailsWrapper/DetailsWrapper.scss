.plateau-data {
  display: flex;

  label {
    display: flex;
    flex-direction: column;
    flex-grow: 1;
    margin: 0 15px;

    span {
      margin-bottom: 15px;

      b {
        font-size: 11px;
      }
    }

    input {
      padding: 15px;
      font-size: 1.1rem;
    }
  }
}

.rover-data {
  height: 280px;
  display: flex;
  flex-direction: column;
  justify-content: space-between;

  .next-steps-resume {
    position: relative;
    top: -10px;
  }
}
