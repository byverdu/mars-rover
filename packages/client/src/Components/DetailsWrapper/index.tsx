import React, { Fragment, useState } from 'react';
import './DetailsWrapper.scss';
import DetailsInput from '../DetailsInput';
import { IPlateauPayload } from '../../Models/Interfaces';

export interface DetailsWrapperProps {
  plateauTitle: string;
  roverTitle: string;
  submitData: (data: IPlateauPayload) => void;
}

const renderPlateauInfo = (plateauWidth, plateauHeight) => {
  if (plateauWidth > 0 || plateauHeight > 0) {
    return (
      <h3>Plateau size {plateauWidth} x {plateauHeight}</h3>
    );
  }

  return null;
}

const renderRoverPosition = (position: string) => {
  if (position.length > 0) {
    return (
      <h3>Initial Rover Position: {position}</h3>
    );
  }

  return null;
}

const renderRoverNextSteps = (steps: string) => {
  if (steps.length > 0) {
    return (
      <h3>Next Steps: {steps}</h3>
    );
  }

  return null;
}

const DetailsWrapper: React.FC<DetailsWrapperProps> = ({
  plateauTitle,
  roverTitle,
  submitData
}) => {

  const [plateauWidth, setPlateauWidth] = useState(0);
  const [plateauHeight, setPlateauHeight] = useState(0);
  const [roverPosition, setRoverPosition] = useState('');
  const [roverNextSteps, setRoverNextSteps] = useState('');

  return (
    <Fragment>
      {renderPlateauInfo(plateauWidth, plateauHeight)}
      {renderRoverPosition(roverPosition)}
      {renderRoverNextSteps(roverNextSteps)}
      <div className="container collapse">
        <DetailsInput title={plateauTitle}>
          <section className="plateau-data">
            <input
              type="number"
              value={plateauWidth}
              placeholder="Set Width"
              onChange={
                (e: React.ChangeEvent<HTMLInputElement>) => setPlateauWidth(Number(e.target.value))
              }
            />
            <input
              type="number"
              value={plateauHeight}
              placeholder="Set Height"
              onChange={
                (e: React.ChangeEvent<HTMLInputElement>) => setPlateauHeight(Number(e.target.value))
              }
            />
          </section>
        </DetailsInput>
      </div>
      <div className="container collapse">
        <DetailsInput title={roverTitle}>
          <section className="rover-data">
            <input
              type="text"
              name="position"
              placeholder="Set Rover Coords"
              value={roverPosition}
              onChange={
                (e: React.ChangeEvent<HTMLInputElement>) => setRoverPosition(e.target.value)
              }
            />
            <input
              type="text"
              name="steps"
              placeholder="Set Rover Steps"
              value={roverNextSteps}
              onChange={
                (e: React.ChangeEvent<HTMLInputElement>) => setRoverNextSteps(e.target.value)
              }
            />
          </section>
        </DetailsInput>
      </div>
      <button onClick={(e) => {
        e.preventDefault();
        submitData({
          plateauSize: `${plateauWidth}x${plateauHeight}`,
          rovers: [
            { position: roverPosition, steps: roverNextSteps }
          ]
        })

      }}>Set Data</button>
    </Fragment>
  );
}

export default DetailsWrapper
