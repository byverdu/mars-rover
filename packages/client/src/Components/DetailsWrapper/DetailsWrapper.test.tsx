import React from 'react';
import DetailsWrapper, { DetailsWrapperProps } from './index';
import { mount, shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

const props: DetailsWrapperProps = {
  plateauTitle: 'Set Plateau Size',
  roverTitle: 'Set Rover Position',
  submitData: () => {}
};

describe('<DetailsWrapper />', () => {
  const wrapper = mount(<DetailsWrapper {...props} />);
  const wrapper_ = shallow(<DetailsWrapper {...props} />);

  it('should create a snapshot', () => {
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('should have a plateauTitle props', () => {
    expect((wrapper.props() as DetailsWrapperProps).plateauTitle).toEqual('Set Plateau Size');
  });

  it('should have a roverTitle props', () => {
    expect((wrapper.props() as DetailsWrapperProps).roverTitle).toEqual('Set Rover Position');
  });

  it('should initially render two DetailsInput components', () => {
    expect(wrapper.find('DetailsInput').length).toEqual(2);
  });

  it('should display the proper text for each DetailsInput component', () => {
    expect(wrapper.find('DetailsInput').first().text()).toEqual(props.plateauTitle);
    expect(wrapper.find('DetailsInput summary').last().text()).toEqual(props.roverTitle);
  });

  it('should set the plateau width properly', () => {
    wrapper_.find(".plateau-data input").first().simulate('change', {target: {value: 9}})
    expect(wrapper_.find("h3").text()).toEqual('Plateau size 9 x 0')
  });

  it('should set the plateau height properly', () => {
    wrapper_.find(".plateau-data input").last().simulate('change', {target: {value: 9}})
    expect(wrapper_.find("h3").text()).toEqual('Plateau size 9 x 9')
  });
});
