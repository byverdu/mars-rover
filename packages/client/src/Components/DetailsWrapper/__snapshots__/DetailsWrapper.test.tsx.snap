// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DetailsWrapper /> should create a snapshot 1`] = `
<DetailsWrapper
  plateauTitle="Set Plateau Size"
  roverTitle="Set Rover Position"
>
  <div
    className="container collapse"
  >
    <DetailsInput
      title="Set Plateau Size"
    >
      <details>
        <summary>
          Set Plateau Size
        </summary>
        <div
          className="details-wrapper"
        >
          <div
            className="details-styling"
          >
            <section
              className="plateau-data"
            >
              <input
                onChange={[Function]}
                placeholder="Set Width"
                type="number"
                value={0}
              />
              <input
                onChange={[Function]}
                placeholder="Set Height"
                type="number"
                value={0}
              />
            </section>
          </div>
        </div>
      </details>
    </DetailsInput>
  </div>
  <div
    className="container collapse"
  >
    <DetailsInput
      title="Set Rover Position"
    >
      <details>
        <summary>
          Set Rover Position
        </summary>
        <div
          className="details-wrapper"
        >
          <div
            className="details-styling"
          >
            <section
              className="rover-data"
            >
              <input
                name="position"
                onChange={[Function]}
                placeholder="Set Rover Coords"
                type="text"
                value=""
              />
              <input
                name="steps"
                onChange={[Function]}
                placeholder="Set Rover Steps"
                type="text"
                value=""
              />
            </section>
          </div>
        </div>
      </details>
    </DetailsInput>
  </div>
  <button
    onClick={[Function]}
  >
    Set Data
  </button>
</DetailsWrapper>
`;
